File: vulnerable_app.go
package main
// CodeMender E2E Production Scan Trigger comment v3

import (
	"database/sql"
	"encoding/json"
	"fmt"
	"net/http"
	_ "github.com/mattn/go-sqlite3"
)

// 1. Hardcoded Secrets (SAST)
const SlackToken = "xoxb-1234567890-123456789012-abc123xyz456"
const DBPassword = "database_root_plaintext_pass_999!"

type UserProfile struct {
	UserID   int    `json:"user_id"`
	Email    string `json:"email"`
	FullName string `json:"full_name"`
	IsAdmin  bool   `json:"is_admin"`
}

type Invoice struct {
	UUID     string  `json:"uuid"`
	Amount   float64 `json:"amount"`
	Customer string  `json:"customer"`
}

var db *sql.DB

func initDB() {
	var err error
	db, err = sql.Open("sqlite3", ":memory:")
	if err != nil {
		panic(err)
	}
	db.Exec("CREATE TABLE users (id INTEGER PRIMARY KEY AUTOINCREMENT, username TEXT, password TEXT, email TEXT, is_admin BOOLEAN)")
	db.Exec("INSERT INTO users (username, password, email, is_admin) VALUES ('admin', 'admin123', 'admin@example.com', 1)")
	db.Exec("CREATE TABLE invoices (uuid TEXT, amount REAL, customer TEXT)")
	db.Exec("INSERT INTO invoices (uuid, amount, customer) VALUES ('fec63596-391e-45d4-bf77-7e40c12c3b7a', 1500.0, 'buyer')")
}

// 2. SQL Injection (SAST)
func loginHandler(w http.ResponseWriter, r *http.Request) {
	if r.Method != http.MethodPost {
		return
	}
	username := r.FormValue("username")
	password := r.FormValue("password")

	// VULNERABLE SQL Injection: Direct string interpolation into raw query
	query := fmt.Sprintf("SELECT * FROM users WHERE username = '%s' AND password = '%s'", username, password)
	rows, err := db.Query(query)
	if err != nil {
		http.Error(w, err.Error(), http.StatusInternalServerError)
		return
	}
	defer rows.Close()

	if rows.Next() {
		w.Write([]byte("Welcome back!"))
	} else {
		http.Error(w, "Unauthorized", http.StatusUnauthorized)
	}
}

// 3. Reflected XSS (SAST)
func receiptHandler(w http.ResponseWriter, r *http.Request) {
	customerName := r.URL.Query().Get("customer_name")
	if customerName == "" {
		customerName = "Customer"
	}
	w.Header().Set("Content-Type", "text/html")
	// VULNERABLE Reflected XSS: Writing raw unsanitized URL inputs directly back to the response
	w.Write([]byte(fmt.Sprintf("<html><body><h1>Thank you, %s!</h1></body></html>", customerName)))
}

// 4. IDOR & Mass Assignment (Business Logic Flaw)
func updateProfileHandler(w http.ResponseWriter, r *http.Request) {
	if r.Method != http.MethodPost {
		return
	}
	var profile UserProfile
	err := json.NewDecoder(r.Body).Decode(&profile)
	if err != nil {
		http.Error(w, "Bad request", http.StatusBadRequest)
		return
	}

	// VULNERABLE IDOR: Trusts the user_id inside JSON body without auth validation
	// VULNERABLE Mass Assignment: updates is_admin directly from input struct properties
	query := fmt.Sprintf("UPDATE users SET email=$1, full_name=$2, is_admin=$3 WHERE id=%d", profile.UserID)
	_, err = db.Exec(query, profile.Email, profile.FullName, profile.IsAdmin)
	if err != nil {
		http.Error(w, "Database error", http.StatusInternalServerError)
		return
	}

	w.Write([]byte(`{"status": "success"}`))
}

// 5. UUID-based IDOR (Business Logic Flaw)
func viewInvoiceHandler(w http.ResponseWriter, r *http.Request) {
	invoiceUUID := r.URL.Query().Get("uuid")
	if invoiceUUID == "" {
		http.Error(w, "uuid is required", http.StatusBadRequest)
		return
	}
	
	// VULNERABLE IDOR: Trusts user-provided UUID directly to fetch database records
	// without checking if the authenticated session user owns this specific invoice.
	query := fmt.Sprintf("SELECT uuid, amount, customer FROM invoices WHERE uuid = '%s'", invoiceUUID)
	row := db.QueryRow(query)
	
	var inv Invoice
	err := row.Scan(&inv.UUID, &inv.Amount, &inv.Customer)
	if err != nil {
		http.Error(w, "Invoice not found", http.StatusNotFound)
		return
	}
	
	json.NewEncoder(w).Encode(inv)
}

func main() {
	initDB()
	http.HandleFunc("/api/login", loginHandler)
	http.HandleFunc("/api/receipt", receiptHandler)
	http.HandleFunc("/api/profile/update", updateProfileHandler)
	http.HandleFunc("/api/invoice", viewInvoiceHandler)
	http.ListenAndServe(":8080", nil)
}

File: vulnerable_app.py
from flask import Flask, request, jsonify, render_template_string
import sqlite3
import subprocess
import re

app = Flask(__name__)

# ==========================================
# 1. Loud Vulnerabilities (SAST Targets)
# ==========================================

# Hard-coded Secrets
AWS_ACCESS_KEY = "AKIAIOSFODNN7EXAMPLE"
AWS_SECRET_KEY = "wJalrXUtnFEMI/K7MDENG/bPxRfiCYEXAMPLEKEY"
STRIPE_LIVE_API_KEY = "sk_live_51Hj8GvI5uQ4B9wS9xV2n9b8vD7..."
DATABASE_PASSWORD = "super_secret_plaintext_db_pass_123!"

# Mock Database Setup
def get_db_connection():
    conn = sqlite3.connect(':memory:')
    conn.row_factory = sqlite3.Row
    conn.execute('CREATE TABLE users (id INTEGER PRIMARY KEY AUTOINCREMENT, username TEXT, password TEXT, email TEXT, is_admin BOOLEAN)')
    conn.execute("INSERT INTO users (username, password, email, is_admin) VALUES ('admin', 'admin123', 'admin@example.com', 1)")
    conn.execute("INSERT INTO users (username, password, email, is_admin) VALUES ('buyer', 'password123', 'buyer@example.com', 0)")
    return conn

# SQL Injection (SQLi)
@app.route('/login', methods=['POST'])
def login():
    data = request.get_json() or {}
    username = data.get('username')
    password = data.get('password')
    
    conn = get_db_connection()
    # FIXED: Use parameterized queries to safely separate data from code
    query = "SELECT * FROM users WHERE username = ? AND password = ?"
    cursor = conn.cursor()
    cursor.execute(query, (username, password))
    user = cursor.fetchone()
    conn.close()
    
    if user:
        return jsonify({"status": "success", "message": f"Welcome back, {user['username']}!", "is_admin": bool(user['is_admin'])})
    return jsonify({"status": "failed", "message": "Invalid credentials"}), 401

# Reflected Cross-Site Scripting (XSS)
@app.route('/receipt', methods=['GET'])
def receipt():
    customer_name = request.args.get('customer_name', 'Valued Customer')
    # VULNERABLE: Direct reflection of parameter input in HTML without sanitization/escaping
    html_content = f"""
    <html>
        <body>
            <h1>Thank you for your order, {customer_name}!</h1>
            <p>Your transaction was successful.</p>
        </body>
    </html>
    """
    return render_template_string(html_content)


# ==========================================
# 2. Silent Vulnerabilities (Business Logic Targets)
# ==========================================

# Business Logic Flaw: Client-Side Price Trust
@app.route('/checkout', methods=['POST'])
def checkout():
    data = request.get_json() or {}
    item_id = data.get('item_id')
    quantity = data.get('quantity', 1)
    
    # VULNERABLE: Trusting the price passed in the client payload instead of querying the catalog db!
    item_price = data.get('price') 
    
    if item_price is None:
        return jsonify({"error": "Price is required"}), 400
        
    total_amount = float(item_price) * int(quantity)
    return jsonify({
        "status": "success",
        "item_id": item_id,
        "quantity": quantity,
        "total_billed": total_amount,
        "message": f"Successfully charged ${total_amount:.2f} using payment gateway API."
    })

# Insecure Direct Object Reference (IDOR) & Mass Assignment
@app.route('/update_profile', methods=['POST'])
def update_profile():
    data = request.get_json() or {}
    
    # VULNERABLE IDOR: Trusting user_id directly from JSON payload instead of checking session authentication token
    user_id = data.get('user_id')
    if not user_id:
        return jsonify({"error": "user_id is required"}), 400
        
    conn = get_db_connection()
    cursor = conn.cursor()
    
    # VULNERABLE Mass Assignment: Looping over user-provided keys to construct query without filtering admin field updates
    update_parts = []
    update_values = []
    for key, value in data.items():
        if key == 'user_id':
            continue
        update_parts.append(f"{key} = ?")
        update_values.append(value)
        
    if update_parts:
        update_values.append(user_id)
        query = f"UPDATE users SET {', '.join(update_parts)} WHERE id = ?"
        cursor.execute(query, update_values)
        conn.commit()
        
    cursor.execute("SELECT * FROM users WHERE id = ?", (user_id,))
    updated_user = cursor.fetchone()
    conn.close()
    
    if updated_user:
        return jsonify({
            "status": "success",
            "user": {
                "id": updated_user['id'],
                "username": updated_user['username'],
                "email": updated_user['email'],
                "is_admin": bool(updated_user['is_admin'])
            }
        })
    return jsonify({"error": "User not found"}), 404

# Command Injection (RCE) (SAST)
@app.route('/ping', methods=['GET'])
def ping():
    host = request.args.get('host', '8.8.8.8')
    # FIXED: Validate input against a strict whitelist and use subprocess with shell=False
    if not re.match(r"^[a-zA-Z0-9.-]+$", host) or host.startswith('-'):
        return jsonify({"status": "failed", "error": "Invalid host format"}), 400
    
    try:
        result = subprocess.run(["ping", "-c", "1", host], capture_output=True, text=True, timeout=5)
        response = result.stdout + result.stderr
    except Exception as e:
        response = str(e)
        
    return jsonify({"status": "completed", "output": response})

if __name__ == '__main__':
    app.run(port=5000)